extensions/amp-story/1.0/amp-story.js

```diff
@@ -2237,13 +2240,44 @@ export class AmpStory extends AMP.BaseElement {
 
     const pagesByDistance = this.getPagesByDistance_();
 
-    this.mutateElement(() => {
+    if (!isExperimentOn(this.win, 'amp-story-load-first-page-only')) {
+      return pagesByDistance.forEach((pageIds, distance) => {
+        pageIds.forEach((pageId) => {
+          const page = this.getPageById(pageId);
+          page.setDistance(distance);
+        });
+      });
+    }
+
+    const preloadAllPages = () => {
       pagesByDistance.forEach((pageIds, distance) => {
         pageIds.forEach((pageId) => {
           const page = this.getPageById(pageId);
           page.setDistance(distance);
         });
       });
+    };
+
+    this.mutateElement(() => {
+      if (prioritizeActivePage) {
+        // Load page with distance 0 first, and then load the other ones.
+        return Promise.all(
+          pagesByDistance[0].map((pageId) => {
+            const page = this.getPageById(pageId);
+            page.setDistance(0);
+            return page.signals().whenSignal(CommonSignals.LOAD_END);
+          })
+        ).then(() => {
+          if (
+            pagesByDistance[0].every(
+              (curr) => this.getPageById(curr).getDistance() == 0
+            )
+          ) {
```

Comment:
This is so much harder to understand than just canceling the promise on navigation. People without advanced context on the feature (basically just you and I for the next couple of months - then we'll forget) won't be able to understand what this does or what it's there for, and we're just making our codebase harder for anyone.